precompiles/erc20/approve.go

```diff
@@ -0,0 +1,284 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package erc20
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/core/vm"
+	auth "github.com/evmos/evmos/v15/precompiles/authorization"
+)
+
+// Approve sets the given amount as the allowance of the spender address over
+// the caller’s tokens. It returns a boolean value indicating whether the
+// operation succeeded and emits the Approval event on success.
+//
+// The Approve method handles 4 cases:
+//  1. no authorization, amount 0 or negative -> return error
+//  2. no authorization, amount positive -> create a new authorization
+//  3. authorization exists, amount 0 or negative -> delete authorization
+//  4. authorization exists, amount positive -> update authorization
+func (p Precompile) Approve(
+	ctx sdk.Context,
+	contract *vm.Contract,
+	stateDB vm.StateDB,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	spender, amount, err := ParseApproveArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	grantee := spender
+	granter := contract.CallerAddress
+
+	// TODO: owner should be the owner of the contract
+	authorization, _, err := auth.CheckAuthzExists(ctx, p.AuthzKeeper, grantee, granter, SendMsgURL)
+	if err != nil {
+		return nil, err
+	}
+	// case 1: authorization doesn't exist
+
+	switch {
+	case authorization == nil && amount != nil && amount.Cmp(common.Big0) <= 0:
+		// case 1: no authorization, amount 0 or negative -> error
+		// TODO: (@fedekunze) check if this is correct by comparing behavior with
+		// regular ERC20
+		err = errors.New("cannot approve non-positive values")
+	case authorization == nil && amount != nil && amount.Cmp(common.Big0) > 0:
+		// case 2: no authorization, amount positive -> create a new authorization
+		err = p.createAuthorization(ctx, grantee, granter, amount)
+	case authorization != nil && amount != nil && amount.Cmp(common.Big0) <= 0:
+		// case 3: authorization exists, amount 0 or negative -> delete authorization
+		err = p.deleteAuthorization(ctx, grantee, granter)
+	case authorization != nil && amount != nil && amount.Cmp(common.Big0) > 0:
```

Comment:
Should't we include also a default that return an error?

Reply:
is there another case other than those? the four cases cover everything